gordon_janitor_gcp/gce_instance_authority.py

```diff
@@ -0,0 +1,145 @@
+
+
+class GCEInstanceAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): Plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to receive instance data
+            messages.
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
+        self.rrset_channel = rrset_channel
+
+    async def get_all_projects(self):
+        """Get all active GCE projects.
+
+        Returns:
+            set: active GCE project IDs.
+        """
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile,
+            scopes=['cloud-platform.read-only'],
+            session=self.session)
+        crm_client = gce_clients.CRMClient(auth_client, self.session)
+
+        active_projects = await crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def get_processed_projects(self):
+        """Get list of GCE project IDs with blacklisted ones filtered out.
+
+        Returns:
+            list: project IDs.
+        """
+        projects = await self.get_all_projects()
+        project_blacklist = set(self.config['project_blacklist'])
+        return sorted(projects - project_blacklist)
+
+    async def get_instance_filter(self):
+        """Get filter string used by instances.aggregatedList endpoint.
+
+        For filter string details, see api endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Returns:
+            str: filter suitable for the 'filter' endpoint parameter.
+        """
+        return ''
+
+    async def get_instances_for_projects(self, projects):
+        """Get list of instance data for each project.
+
+        Args:
+            projects (list): list of project names.
+        Yields:
+            list: list of dicts that contain instance information.
+        """
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        metadata_blacklist = self.config.get('metadata_blacklist', {})
+
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile, scopes=['cloud-platform'], session=self.session)
+
+        gcp_client = gce_clients.GCPClient(
+            auth_client,
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+        instance_filter = await self.get_instance_filter()
+
+        for project in projects:
+            yield await gcp_client.list_instances(
+                project, instance_filter=instance_filter)
+
+    async def get_msg_generator(self):
+        """Get message generator to translate instance data into channel msgs.
+
+        Returns:
+            obj: obj that creates messages for the channel.
+        """
+        return msg_generators.ARecordMsgGenerator(self.config['zone'])
```

Comment:
A couple of things: this is an unhandled `KeyError` if `'zone'` is not in config. 

And, following the `msg_generator` logic, the zone in config is the DNS zone. But I think I am not understanding how the instances are actually filtered for the given zone. 

That also got me thinking (and I could be completely misunderstanding your code - I'm so sorry if I am): if we're configuring for one DNS zone, then we'll need multiple janitors to run, one for each DNS zone. Perhaps there can be a list of zones provided to the authority, so then all instances of all projects are polled, then divvied up into the batched messages. 

I'm hoping/thinking that the list of zones could almost be a global gcp plugin janitor config item, but it might be Authority client specific, or Reconciler specific (i.e. Cloud DNS refers to zones as `managedZones` - where should that go in config? does this name even matter if the DNS client knows what we're talking about when we just say `zones`?

Reply:
Also, I assume `PTR` and `CNAME` records are generic need to address. Half-baked/jet lagged idea: perhaps this can be configurable, e.g. `record_types_to_generate` or `record_types_to_reconcile` or just `record_types` on the global plugin level:

```toml
[gcp]
record_types = ["A", "CNAME", "PTR"]
```

What do you think?